plugins/module_utils/user.py

```diff
@@ -924,6 +967,51 @@ def limit_resources(module, cursor, user, host, resource_limits, check_mode):
     return True
 
 
+def get_attribute_support(cursor):
+    """Checks if the MySQL server supports user attributes.
+
+    Args:
+        cursor (cursor): DB driver cursor object.
+    Returns:
+        True if attributes are supported, False if they are not.
+    """
+
+    try:
+        # information_schema.tables does not hold the tables within information_schema itself
+        cursor.execute("SELECT attribute FROM INFORMATION_SCHEMA.USER_ATTRIBUTES LIMIT 0")
+        cursor.fetchone()
+    except mysql_driver.Error:
+        return False
+
+    return True
+
+
+def attributes_get(cursor, user, host):
+    """Get attributes for a given user.
+
+    Args:
+        cursor (cursor): DB driver cursor object.
+        user (str): User name.
+        host (str): User host name.
+
+    Returns:
+        None if the user does not exist, otherwise a dict of attributes set on the user
+    """
+    cursor.execute("SELECT attribute FROM INFORMATION_SCHEMA.USER_ATTRIBUTES WHERE user = %s AND host = %s", (user, host))
+
+    r = cursor.fetchone()
+
+    if r:
+        attributes = r[0]
+        # convert JSON string stored in row into a dict - mysql enforces that user_attributes entires are in JSON format
+        if attributes:
+            return json.loads(attributes)
+        else:
+            return {}
+
+    return None
```

Comment:
Looks like this will return `None` rather than an empty dict if the user exists but does not have any attributes set - are you recommending returning `None` over `{}` in that scenario?

Reply:
yep

Reply:
That should be fine, but there are a few things to keep in mind about `None` vs `{}` when it comes to attributes... if a `CREATE USER` is ran without the `ATTRIBUTES '{}'` option, the user's attributes are stored as `NULL` until attributes are set, and if all attributes are deleted, attributes will then be stored as `{}` and cannot be changed back to `NULL`  (as far as I can tell):

```
mysql> CREATE USER 'testuser'@'localhost';
Query OK, 0 rows affected (0.00 sec)

mysql> SELECT ATTRIBUTE from INFORMATION_SCHEMA.USER_ATTRIBUTES WHERE user = 'testuser' and host = 'localhost';
+-----------+
| ATTRIBUTE |
+-----------+
| NULL      |
+-----------+
1 row in set (0.00 sec)

mysql> ALTER USER 'testuser'@'localhost' ATTRIBUTE '{ "foo": "bar" }';
Query OK, 0 rows affected (0.00 sec)

mysql> SELECT ATTRIBUTE from INFORMATION_SCHEMA.USER_ATTRIBUTES WHERE user = 'testuser' and host = 'localhost';
+----------------+
| ATTRIBUTE      |
+----------------+
| {"foo": "bar"} |
+----------------+
1 row in set (0.00 sec)

mysql> ALTER USER 'testuser'@'localhost' ATTRIBUTE '{ "foo": null }';
Query OK, 0 rows affected (0.00 sec)

mysql> SELECT ATTRIBUTE from INFORMATION_SCHEMA.USER_ATTRIBUTES WHERE user = 'testuser' and host = 'localhost';
+-----------+
| ATTRIBUTE |
+-----------+
| {}        |
+-----------+
1 row in set (0.00 sec)

mysql>
```

I think there are a couple different options:

1. Represent no attributes as `{}` even if the `ATTRIBUTE` database value is `NULL`
    1. This is what I was doing originally; I chose this method so that someone using `register: foo` with `mysql_user` could still treat `foo.attributes` as a dictionary even if the `ATTRIBUTE` value is `NULL`. It also makes the code a bit cleaner if `attributes_get` always returns a dict for users that exist.
    1. With this method, we could also explicitly run the `CREATE USER` query in `user_add` with `ATTRIBUTES {}` even if no attributes are specified, so that the `ATTRIBUTE` database value will always be `{}` for ansible-created users on servers with attribute support.
1. Represent no attributes as `None` even if the `ATTRIBUTE` database value is `{}` (I think this is what you were proposing)
1. Represent no attributes as either `None` or `{}` depending on the database value

Option 2 is how this feature behaves as of my latest commit (I also wrote tests to assert the attributes return value for users without attributes). I think any of these options are fine depending on ansible best practices. I can stick with option 2 if you think it's preferred, I just wanted to make sure this caveat was documented in this PR.